reladomographql/src/main/java/com/gs/reladomo/graphql/SDLGenerator.java

```diff
@@ -0,0 +1,319 @@
+ KIND, either express or implied.  See the License for the
+ specific language governing permissions and limitations
+ under the License.
+ */
+
+package com.gs.reladomo.graphql;
+
+import com.gs.fw.common.mithra.MithraManagerProvider;
+import com.gs.fw.common.mithra.attribute.*;
+import com.gs.fw.common.mithra.finder.AbstractRelatedFinder;
+import com.gs.fw.common.mithra.finder.Mapper;
+import com.gs.fw.common.mithra.finder.RelatedFinder;
+import com.gs.fw.common.mithra.transaction.MithraTransactionalResource;
+import com.gs.fw.common.mithra.util.MithraRuntimeCacheController;
+import org.eclipse.collections.impl.set.mutable.UnifiedSet;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.File;
+import java.io.FileWriter;
+import java.io.IOException;
+import java.io.Writer;
+import java.time.LocalDateTime;
+import java.util.Set;
+import java.util.function.BiFunction;
+
+public class SDLGenerator
+{
+    private static final Logger LOGGER = LoggerFactory.getLogger (SDLGenerator.class);
+    private static boolean withComments = false;
+
+    public void generate (final String filename)
+    {
+        try (final Writer out = new FileWriter (filename))
+        {
+            if (withComments) out.write ("# generated on " + LocalDateTime.now () + "\n");
+            this.generate (out);
+
+        } catch (final Exception e)
+        {
+            throw new RuntimeException ("failed to write into " + new File (filename).getAbsolutePath (), e);
+        }
+
+        LOGGER.info ("The complete schema was written to " + new File (filename).getAbsolutePath ());
+    }
+
+    protected void generate (final Writer out) throws IOException
+    {
+        this.writeMutationBlock (out);
+
+        final Set<String> declaredTypes = this.writeQueryBlock (out);
+
+        for (final MithraRuntimeCacheController controller : MithraManagerProvider.getMithraManager ().getRuntimeCacheControllerSet ())
+        {
+            String name = controller.getClassName ();
+            name = StringKit.decapitalize (name, name.lastIndexOf ('.') + 1);
+            name = StringKit.capitalize (name);
+            final RelatedFinder finder = controller.getFinderInstance ();
+
+            out.write ("#################### " + controller.getClassName () + " ##################\n");
+
+            generateType (out, declaredTypes, name, finder);
+            generateAggregate (out, declaredTypes, name, finder);
+            generateInput (out, name, finder);
+            generateOrderBy (out, name, finder);
+            generateFinder (out, declaredTypes, name, finder);
+            generateTupleInput (out, name, finder);
+        }
+    }
+
+    private void generateType (final Writer out, final Set<String> declaredTypes, final String name, final RelatedFinder finder) throws IOException
+    {
+        out.write ("type " + name + "\n{\n");
+        for (final Attribute attr : finder.getPersistentAttributes ())

```

Comment:
what about source attributes?

Reply:
added support for the source attributes on queries and mutations